### PR TITLE
Feature/50 support merge statement

### LINKED ISSUE #46: Implement missed expressions and clauses for integration sql-statement-builder in virtual-schemas
### Problem: 
We want to use sql-statement-builder for rendering sql statements in virtual-schemas project. 
Unfortunate, sql-statement-builder doesn't support all necessary functionalities. 

### Solution:
Add missed expressions and clauses.
For example: 
-GROUP BY
-ORDER BY
-HAVING
etc.

### LINKED ISSUE #50: Support MERGE Statement
# Situation

The [MERGE](https://docs.exasol.com/sql/merge.htm) statement provides a combination of `INSERT`, `UPDATE` and `DELETE`.

It is especially useful in cases where new data should be combined with data that already exists in a table.

# Acceptance Criteria
1. `MERGE` supported without `WHEN`

### LINKED ISSUE #52: Support `WHERE` in `MERGE`
## Situation

Basic `MERGE` support was introduced with #51. Support for `WHERE` is still missing in the `THEN UPDATE`, `THEN DELETE` and `THEN INSERT` clauses.

## Acceptance Criteria

1. `WHERE` supported in `THEN UPDATE`, `THEN DELETE` and `THEN INSERT`.

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #52.
Closes #46.

### CODE REVIEW

File: src/main/java/com/exasol/sql/dml/merge/Merge.java [hunk 1]
```diff
@@ -71,42 +89,60 @@ public Merge on(final BooleanExpression condition) {
         return this;
     }
 
+    /**
+     * Get the merge condition.
+     *
+     * @return criteria that must be met for the rows in source and destination to be considered a match.
+     */
+    public BooleanExpression getCondition() {
+        return this.condition;
```
Comment (morazow) [inline src/main/java/com/exasol/sql/dml/merge/Merge.java]: Seems like this `condition` should returned from `on` object.

File: src/main/java/com/exasol/sql/dml/merge/Merge.java [hunk 2]
```diff
@@ -71,42 +89,60 @@ public Merge on(final BooleanExpression condition) {
...
+    }
+
+    /**
+     * Check if the {@code ON} clause exists.
+     *
+     * @return {@code true} if the {@code ON} clause exists.
+     */
+    protected boolean hasCondition() {
+        return this.on != null;
```
Comment (morazow) [inline src/main/java/com/exasol/sql/dml/merge/Merge.java]: Similar to above comment, check if `this.on.condition` is not null?

Comment (morazow) [reply src/main/java/com/exasol/sql/dml/merge/Merge.java]: After reading the javadoc, maybe you wan to check if `on` clause exists, then rename method to hasOn.

File: src/main/java/com/exasol/sql/dml/merge/MergeMethodDefinition.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+     * @return parent {@code MERGE} statement
+     */
+    public Merge where(final BooleanExpression expression) {
+        final Merge merge = (Merge) this.getRoot();
+        this.where = new WhereClause(merge, expression);
+        return merge;
+    }
+
+    /**
+     * Get the {@code WHERE} clause of the update definition.
```
Comment (morazow) [inline src/main/java/com/exasol/sql/dml/merge/MergeMethodDefinition.java]: `.. clause of update or delete definition.`

Or remove `update` all together.

File: src/main/java/com/exasol/sql/dml/merge/MergeVisitor.java [hunk 4]
```diff
@@ -93,8 +94,15 @@
...
-     * 
+     *
      * @param field field reference
      */
     public void visit(final Field field);
+
+    /**
+     * Visit a {@code WHERE} clause.
+     *
+     * @param whereClause the {@code WHERE clause to be visited}
```
Comment (morazow) [inline src/main/java/com/exasol/sql/dml/merge/MergeVisitor.java]:
```suggestion
     * @param whereClause the {@code WHERE} clause to be visited
```

File: src/main/java/com/exasol/sql/dml/merge/rendering/MergeRenderer.java [hunk 5]
```diff
@@ -84,6 +85,7 @@ public void visit(final NotMatchedClause notMatchedClause) {
     }
 
     @Override
+
```
Comment (morazow) [inline src/main/java/com/exasol/sql/dml/merge/rendering/MergeRenderer.java]:
```suggestion
```

File: src/main/java/com/exasol/sql/dml/merge/Merge.java (hunk 1 above)
Comment (redcatbear) [reply src/main/java/com/exasol/sql/dml/merge/Merge.java]: Fixed. Also fixed names.

File: src/main/java/com/exasol/sql/dml/merge/Merge.java (hunk 2 above)
Comment (redcatbear) [reply src/main/java/com/exasol/sql/dml/merge/Merge.java]: Fixed.

File: src/main/java/com/exasol/sql/dml/merge/MergeMethodDefinition.java (hunk 3 above)
Comment (redcatbear) [reply src/main/java/com/exasol/sql/dml/merge/MergeMethodDefinition.java]: Fixed.

File: src/main/java/com/exasol/sql/dml/merge/MergeVisitor.java (hunk 4 above)
Comment (redcatbear) [reply src/main/java/com/exasol/sql/dml/merge/MergeVisitor.java]: Fixed.

File: src/main/java/com/exasol/sql/dml/merge/rendering/MergeRenderer.java (hunk 5 above)
Comment (redcatbear) [reply src/main/java/com/exasol/sql/dml/merge/rendering/MergeRenderer.java]: Fixed.

File: src/main/java/com/exasol/sql/dml/merge/MergeMethodDefinition.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * Abstract base class for merge method definitions like {@code WHEN MATCHED THEN UPDATE}.
+ */
+public abstract class MergeMethodDefinition extends AbstractFragment {
+    protected WhereClause where = null;
+
+    /**
+     * Create the abstract base for a merge method definition.
+     *
+     * @param root root {@code MEREG} statement
```
Comment (morazow) [inline src/main/java/com/exasol/sql/dml/merge/MergeMethodDefinition.java]:
```suggestion
     * @param root root {@code MERGE} statement
```

**Pull request merged**